### PR TITLE
add auto repay contract with tests and cli

### PR DESCRIPTION
**Pull request by alexandr-martirosyan**

### CODE REVIEW

File: contracts/adapters/AutoRepay.sol [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+  uint256 public constant FEE = 10;
+  uint256 public constant FEE_DECIMALS = 10000;
+
+  constructor(
+    ILendingPoolAddressesProvider addressesProvider,
+    IUniswapV2Router02 uniswapRouter,
+    address wethAddress
+  ) public BaseUniswapAdapter(addressesProvider, uniswapRouter, wethAddress) {}
+
+  function whitelistAddress(address userAddress) public onlyOwner returns (bool) {
```
Comment (ytumanov) [inline contracts/adapters/AutoRepay.sol:49]: external

File: contracts/adapters/AutoRepay.sol [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+  function getWitelistedAddresses() public view returns (address[] memory) {
+    uint256 length = _whitelistedAddresses.length();
+    address[] memory addresses = new address[](length);
+    for (uint256 i = 0; i < length; i++) {
+      addresses[i] = _whitelistedAddresses.at(i);
+    }
+    return addresses;
+  }
+
+  function setMinMaxHealthFactor(uint256 minHealthFactor, uint256 maxHealthFactor) public {
```
Comment (ytumanov) [inline contracts/adapters/AutoRepay.sol:70]: should caller be whitelisted?

File: contracts/adapters/AutoRepay.sol [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+          ? repayParams.debtRepayAmount
+          : currentDebt;
+      }
+      uint256 maxCollateralToSwap = repayParams.collateralAmount;
+      if (amountToRepay < repayParams.debtRepayAmount) {
+        maxCollateralToSwap = maxCollateralToSwap.mul(amountToRepay).div(
+          repayParams.debtRepayAmount
+        );
+      }
+      repayParams.collateralAmount = maxCollateralToSwap;
```
Comment (ytumanov) [inline contracts/adapters/AutoRepay.sol:201]:
```suggestion
      if (amountToRepay < repayParams.debtRepayAmount) {
        repayParams.collateralAmount = maxCollateralToSwap.mul(amountToRepay).div(
          repayParams.debtRepayAmount
        );
      }
```

File: contracts/adapters/AutoRepay.sol [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        // Pull aTokens from user
+        _transferATokenToContractAddress(
+          collateralATokenAddress,
+          repayParams.user,
+          amounts0.add(feeAmount),
+          permitSignature
+        );
+        IERC20(collateralATokenAddress).safeTransfer(caller, feeAmount);
+        LENDING_POOL.withdraw(repayParams.collateralAsset, amounts0, address(this));
+      }
```
Comment (ytumanov) [inline contracts/adapters/AutoRepay.sol:255]:
```suggestion
  _transferATokenToContractAddress(
          collateralATokenAddress,
          repayParams.user,
          amounts0.add(feeAmount),
          permitSignature
        );
        IERC20(collateralATokenAddress).safeTransfer(caller, feeAmount);
      if (!repayParams.useATokenAsFrom) {
       LENDING_POOL.withdraw(repayParams.collateralAsset, amounts0, address(this));
      }
```

File: contracts/adapters/AutoRepay.sol [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+      uint256 contractBalanceBefore = IERC20(collateralATokenAddress).balanceOf(address(this));
+      LENDING_POOL.withdraw(
+        repayParams.collateralAsset,
+        repayParams.debtRepayAmount.add(premium),
+        address(this)
+      );
+      uint256 contractBalanceAfter = IERC20(collateralATokenAddress).balanceOf(address(this));
+      IERC20(collateralATokenAddress).safeTransfer(
+        caller,
+        aTokenTransferAmount - (contractBalanceBefore - contractBalanceAfter)
```
Comment (ytumanov) [inline contracts/adapters/AutoRepay.sol:294]: Probably not needed, but maybe better to use safeMath like in other places

**Review by ytumanov — CHANGES_REQUESTED** [review]

File: cli.js [hunk 6]
```diff
@@ -904,9 +912,161 @@ async function execute(network, action, ...params) {
...
+        .call();
+      maxCollateralAmount = BN(amounts[0])
+        .plus(BN(amounts[0]).multipliedBy(1).dividedBy(1000))
+        .toFixed(0); // 0.1% slippage
+    }
+    const feeAmount = BN(maxCollateralAmount).multipliedBy(10).dividedBy(10000);
+
+    console.log(`Checking mToken ${mToken.options.address} for approval`);
+    if (
+      BN(await mToken.methods.allowance(user, autoRepay.options.address).call()).lt(
```
Comment (ytumanov) [inline cli.js:979]: Maybe send approve tx if there is not enough allowance, just do the same approve like here in other actions

File: cli.js [hunk 7]
```diff
@@ -904,9 +912,161 @@ async function execute(network, action, ...params) {
...
+    const mToken = new eth.Contract(MToken, reserveCollateralToken.aTokenAddress);
+    const reserveDebtToken = await dataProvider.methods
+      .getReserveTokensAddresses(debtAsset.options.address)
+      .call();
+
+    let maxCollateralAmount = 0;
+    if (collateralAsset != debtAsset) {
+      const uniswap = new kit.web3.eth.Contract(
+        Uniswap,
+        '0xe3d8bd6aed4f159bc8000a9cd47cffdb95f96121'
```
Comment (ytumanov) [inline cli.js:960]: Move this to the network switch case

**Review by ytumanov — CHANGES_REQUESTED** [review]

File: cli.js [hunk 8]
```diff
@@ -904,9 +912,161 @@ async function execute(network, action, ...params) {
     );
     return;
   }
-  
+
+  if (action == 'auto-repay') {
```
Comment (x-moola) [inline cli.js:916]: I think it's usually recommended to always use `===` and `!==` in javascript

File: cli.js [hunk 9]
```diff
@@ -904,9 +912,161 @@ async function execute(network, action, ...params) {
...
+        collateralAsset: collateralAsset.options.address,
+        debtAsset: debtAsset.options.address,
+        collateralAmount: maxCollateralAmount.toString(0),
+        debtRepayAmount: repayAmount.toFixed(0),
+        rateMode: rateMode,
+        useEthPath: false,
+        useATokenAsFrom: useATokenAsFrom,
+        useATokenAsTo: useATokenAsTo,
+        useFlashloan: useFlashloan,
+      },
```
Comment (x-moola) [inline cli.js:999]: Can be updated to use the shorthand, something like:
```
      {
        user,
        collateralAsset: collateralAsset.options.address,
        debtAsset: debtAsset.options.address,
        collateralAmount: maxCollateralAmount.toString(0),
        debtRepayAmount: repayAmount.toFixed(0),
        rateMode,
        useEthPath: false,
        useATokenAsFrom,
        useATokenAsTo,
        useFlashloan,
      },
```
to make it cleaner and easier to read

File: cli.js [hunk 10]
```diff
@@ -904,9 +912,161 @@ async function execute(network, action, ...params) {
...
+      pk = params[3];
+      if (!pk) {
+        console.error('Missing private key');
+        return;
+      }
+      kit.addAccount(pk);
+    }
+
+    if (!isNumeric(params[1])) return;
+    if (!isNumeric(params[2])) return;
```
Comment (x-moola) [inline cli.js:1047]: Will it be better to add and display some helpful error messages here?

File: contracts/adapters/AutoRepay.sol [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+
+  /**
+   * @dev Uses the received funds from the flash loan to repay a debt on the protocol on behalf of the user. Then pulls
+   * the collateral from the user and swaps it to the debt asset to repay the flash loan.
+   * The user should give this contract allowance to pull the ATokens in order to withdraw the underlying asset, swap it
+   * and repay the flash loan.
+   * Supports only one asset on the flash loan.
+   * @param assets Address of debt asset
+   * @param amounts Amount of the debt to be repaid
+   * @param premiums Fee of the flash loan
```
Comment (x-moola) [inline contracts/adapters/AutoRepay.sol:106]: It seems the documentation for param `address initiator` is missing

File: contracts/adapters/AutoRepay.sol [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+    _doSwapAndPullWithFee(repayParams, permitSignature, caller, premiums[0]);
+
+    // Repay flashloan. Approves for 0 first to comply with tokens that implement the anti frontrunning approval fix.
+    IERC20(repayParams.debtAsset).safeApprove(address(LENDING_POOL), 0);
+    IERC20(repayParams.debtAsset).safeApprove(address(LENDING_POOL), amounts[0].add(premiums[0]));
+
+    return true;
+  }
+
+  function increaseHealthFactor(
```
Comment (x-moola) [inline contracts/adapters/AutoRepay.sol:167]: Is it possible to add documentations for new functions?

File: contracts/adapters/AutoRepay.sol [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+      repaidAmount = repaidAmount.sub(IERC20(repayParams.debtAsset).balanceOf(address(this)));
+
+      uint256 maxCollateralToSwap = repayParams.collateralAmount;
+      if (repaidAmount < repayParams.debtRepayAmount) {
+        maxCollateralToSwap = maxCollateralToSwap.mul(repaidAmount).div(
+          repayParams.debtRepayAmount
+        );
+      }
+
+      repayParams.collateralAmount = maxCollateralToSwap;
```
[the remark this answers is not in the document]
Comment (lastperson) [reply contracts/adapters/AutoRepay.sol:154]: Don't even need to introduce local var maxCollateralToSwap, just keep using `repayParams.collateralAmount`.

File: contracts/adapters/AutoRepay.sol (hunk 3 above)
Comment (lastperson) [reply contracts/adapters/AutoRepay.sol:201]: again, no need to introduce `maxCollateralToSwap` at all.

File: contracts/adapters/AutoRepay.sol (hunk 5 above)
Comment (lastperson) [reply contracts/adapters/AutoRepay.sol:294]: Why not withdraw everything that is left?

File: cli.js (hunk 6 above)
Comment (alexandr-martirosyan) [reply cli.js:979]: We can't approve because we don't have the user's private key.

File: contracts/adapters/AutoRepay.sol [hunk 14]
```diff
@@ -15,6 +15,20 @@ contract AutoRepay is BaseUniswapAdapter {
...
+   * @dev struct RepayParams
+   *
+   * @param user Address of user
+   * @param colalteralAsset Address of asset to be swapped
+   * @param debtAsset Address of debt asset
+   * @param collateralAmount Amount of the collateral to be swapped
+   * @param debtRepayAmount Amount of the debt to be repaid
+   * @param rateMode Rate mode of the debt to be repaid
+   * @param useEthPath Use Eth in swap path
+   * @param useATokenAsFrom Use aToken as form in swap
```
Comment (lastperson) [inline contracts/adapters/AutoRepay.sol:28]:
```suggestion
   * @param useATokenAsFrom Use aToken as from in swap
```

File: contracts/adapters/AutoRepay.sol [hunk 15]
```diff
@@ -104,15 +118,11 @@ contract AutoRepay is BaseUniswapAdapter {
    * @param assets Address of debt asset
    * @param amounts Amount of the debt to be repaid
    * @param premiums Fee of the flash loan
+   * @param initiator Address of the caller
```
Comment (lastperson) [inline contracts/adapters/AutoRepay.sol:121]:
```suggestion
   * @param initiator Address of the flashloan caller
```

File: contracts/adapters/AutoRepay.sol [hunk 16]
```diff
@@ -215,6 +228,14 @@ contract AutoRepay is BaseUniswapAdapter {
     _checkHealthFactorInRange(repayParams.user);
   }
 
+  /**
+   * @dev does swap and pull and takes 0.1% fee for caller
```
Comment (lastperson) [inline contracts/adapters/AutoRepay.sol:232]: The concepts of "swap" and "pull" are not clear. Need to describe the movement of tokens and what is swapped.

File: contracts/adapters/AutoRepay.sol [hunk 17]
```diff
@@ -282,20 +294,23 @@ contract AutoRepay is BaseUniswapAdapter {
         aTokenTransferAmount,
         permitSignature
       );
-      uint256 contractBalanceBefore = IERC20(collateralATokenAddress).balanceOf(address(this));
+      // uint256 contractBalanceBefore = IERC20(collateralATokenAddress).balanceOf(address(this));
```
Comment (lastperson) [inline contracts/adapters/AutoRepay.sol:297]:
```suggestion
```

File: contracts/adapters/AutoRepay.sol [hunk 18]
```diff
@@ -282,20 +294,23 @@ contract AutoRepay is BaseUniswapAdapter {
...
       LENDING_POOL.withdraw(
         repayParams.collateralAsset,
         repayParams.debtRepayAmount.add(premium),
         address(this)
       );
-      uint256 contractBalanceAfter = IERC20(collateralATokenAddress).balanceOf(address(this));
+      // uint256 contractBalanceAfter = IERC20(collateralATokenAddress).balanceOf(address(this));
```
Comment (lastperson) [inline contracts/adapters/AutoRepay.sol:303]:
```suggestion
```

File: contracts/adapters/AutoRepay.sol [hunk 19]
```diff
@@ -282,20 +294,23 @@ contract AutoRepay is BaseUniswapAdapter {
...
         caller,
-        aTokenTransferAmount - (contractBalanceBefore - contractBalanceAfter)
+        IERC20(collateralATokenAddress).balanceOf(address(this))
       );
     }
   }
 
+  /**
+   * @dev decodes the params and returns them
+   */
```
Comment (lastperson) [inline contracts/adapters/AutoRepay.sol:313]:
```suggestion
```
Docs for the sake of docs are pointless. If the function is simple enough that commentary doesn't help to explain it, then no need for commentary.

**Review by lastperson — CHANGES_REQUESTED** [review]
Cleanup and CLI addresses update.

**Review by lastperson — APPROVED** [review]

**Pull request merged**